File: src/test/java/domain/TryCountTest.java
package domain;

import static Message.ErrorMessage.TRY_COUNT_FORMAT_ERROR_MESSAGE;
import static Message.ErrorMessage.TRY_COUNT_NEGATIVE_ERROR_MESSAGE;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.junit.jupiter.api.Assertions.assertAll;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

class TryCountTest {


    @ParameterizedTest
    @ValueSource(strings = {"0", "1", "2", "3", "2147483647"})
    void 시도횟수를_생성한다(String value) {
        var tryCount = TryCount.from(value);

        assertThat(tryCount).isInstanceOf(TryCount.class);
    }

    @ParameterizedTest
    @ValueSource(strings = {"a", "b", "c", "1a", "a2", "1a1", "2147483648"})
    void 시도횟수를_생성할_때_INT_형이_아니면_예외가_발생한다(String value) {
        assertThatThrownBy(() -> TryCount.from(value))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessage(TRY_COUNT_FORMAT_ERROR_MESSAGE.getValue());
    }

    @ParameterizedTest
    @ValueSource(strings = {"-1", "-2", "-100", "-1000"})
    void 시도횟수를_생성할_때_음수이면_예외가_발생한다(String value) {
        assertThatThrownBy(() -> TryCount.from(value))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessage(TRY_COUNT_NEGATIVE_ERROR_MESSAGE.getValue());
    }

    @Test
    void 시도횟수가_0이면_TRUE_를_반환한다() {
        TryCount tryCount = TryCount.from("0");

        assertThat(tryCount.isZero()).isTrue();
    }

    @Test
    void 시도횟수가_0이_아니면_TRUE_를_반환한다() {
        TryCount tryCount = TryCount.from("1");

        assertThat(tryCount.isZero()).isFalse();
    }

    @Test
    void 시도횟수가_감소한다() {
        TryCount tryCount = TryCount.from("2");

        TryCount oneTryCount = tryCount.decrease();
        TryCount zeroTryCount = tryCount.decrease();

        assertAll(
                () -> assertThat(oneTryCount.isZero()).isFalse(),
                () -> assertThat(zeroTryCount.isZero()).isTrue()
        );
    }
}

File: src/test/java/domain/TryCountTest.kt
package domain

import domain.TryCount.Companion.from
import message.ErrorMessage
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Assertions.assertAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.function.Executable
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource

internal class TryCountTest {

    @ParameterizedTest
    @ValueSource(strings = ["0", "1", "2", "3", "2147483647"])
    fun `시도횟수를 생성한다`(value: String) {
        val tryCount = from(value)

        assertThat(tryCount).isInstanceOf(TryCount::class.java)
    }

    @ParameterizedTest
    @ValueSource(strings = ["a", "b", "c", "1a", "a2", "1a1", "2147483648"])
    fun `시도횟수를 생성할 때 INT형이 아니면 예외가발생한다`(value: String) {
        assertThatThrownBy { from(value) }
            .isInstanceOf(IllegalArgumentException::class.java)
            .hasMessage(ErrorMessage.TRY_COUNT_FORMAT_ERROR_MESSAGE.value)
    }

    @ParameterizedTest
    @ValueSource(strings = ["-1", "-2", "-100", "-1000"])
    fun `시도횟수를 생성할 때 음수이면 예외가 발생한다`(value: String) {
        assertThatThrownBy { from(value) }
            .isInstanceOf(IllegalArgumentException::class.java)
            .hasMessage(ErrorMessage.TRY_COUNT_NEGATIVE_ERROR_MESSAGE.value)
    }

    @Test
    fun `시도횟수가 0이면 TRUE를 반환한다`() {
        val tryCount = from("0")

        assertThat(tryCount.isZero).isTrue
    }

    @Test
    fun `시도횟수가 0이 아니면 TRUE를 반환한다`() {
        val tryCount = from("1")

        assertThat(tryCount.isZero).isFalse
    }

    @Test
    fun `시도횟수가 감소한다`() {
        val tryCount = from("2")

        val oneTryCount = tryCount.decrease()
        val zeroTryCount = oneTryCount.decrease()

        assertAll(
            Executable { assertThat(oneTryCount.isZero).isFalse() },
            Executable { assertThat(zeroTryCount.isZero).isTrue() }
        )
    }
}
